internal/controller/standalone_pgadmin/users_test.go

```diff
@@ -0,0 +1,124 @@
+
+package standalone_pgadmin
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"testing"
+
+	"gotest.tools/v3/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/crunchydata/postgres-operator/internal/naming"
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
+)
+
+func TestReconcilePGAdminUsers(t *testing.T) {
+	ctx := context.Background()
+
+	pgadmin := &v1beta1.PGAdmin{}
+	pgadmin.Namespace = "ns1"
+	pgadmin.Name = "pgadmin1"
+	pgadmin.UID = "123"
+
+	t.Run("NoUsersInSpec", func(t *testing.T) {
+		r := new(PGAdminReconciler)
+		r.Client = fake.NewClientBuilder().Build()
+		assert.NilError(t, r.reconcilePGAdminUsers(ctx, pgadmin))
+	})
+
+	pgadmin.Spec.Users = []v1beta1.PGAdminUser{
+		{
+			Username: "testuser",
+			IsAdmin:  true,
+		},
+	}
+
+	t.Run("NoPods", func(t *testing.T) {
+		r := new(PGAdminReconciler)
+		r.Client = fake.NewClientBuilder().Build()
+		assert.NilError(t, r.reconcilePGAdminUsers(ctx, pgadmin))
+	})
+
+	// Pod in the namespace
+	pod := corev1.Pod{}
+	pod.Namespace = pgadmin.Namespace
+	pod.Name = fmt.Sprintf("pgadmin-%s-0", pgadmin.UID)
+	// pod.Labels = naming.StandalonePGAdminCommonLabels(pgadmin)
+
+	t.Run("ContainerNotRunning", func(t *testing.T) {
+		pod := pod.DeepCopy()
+
+		pod.DeletionTimestamp = nil
+		pod.Status.ContainerStatuses = nil
+
+		r := new(PGAdminReconciler)
+		r.Client = fake.NewClientBuilder().WithObjects(pod).Build()
+
+		assert.NilError(t, r.reconcilePGAdminUsers(ctx, pgadmin))
+	})
+
+	t.Run("PodTerminating", func(t *testing.T) {
+		pod := pod.DeepCopy()
+
+		pod.DeletionTimestamp = new(metav1.Time)
+		*pod.DeletionTimestamp = metav1.Now()
+		pod.Status.ContainerStatuses =
+			[]corev1.ContainerStatus{{Name: naming.ContainerPGAdmin}}
+		pod.Status.ContainerStatuses[0].State.Running =
+			new(corev1.ContainerStateRunning)
+
+		r := new(PGAdminReconciler)
+		r.Client = fake.NewClientBuilder().WithObjects(pod).Build()
+
+		assert.NilError(t, r.reconcilePGAdminUsers(ctx, pgadmin))
+	})
+
+	t.Run("PodHealthy", func(t *testing.T) {
+		pgadmin := pgadmin.DeepCopy()
+		pod := pod.DeepCopy()
+
+		pod.DeletionTimestamp = nil
+		pod.Status.ContainerStatuses =
+			[]corev1.ContainerStatus{{Name: naming.ContainerPGAdmin}}
+		pod.Status.ContainerStatuses[0].State.Running =
+			new(corev1.ContainerStateRunning)
+
+		r := new(PGAdminReconciler)
+		r.Client = fake.NewClientBuilder().WithObjects(pod).Build()
+
+		calls := 0
+		r.PodExec = func(
+			namespace, pod, container string,
+			stdin io.Reader, stdout, stderr io.Writer, command ...string,
+		) error {
+			calls++
+
+			assert.Equal(t, pod, "pgadmin-123-0")
+			assert.Equal(t, namespace, pgadmin.Namespace)
+			assert.Equal(t, container, naming.ContainerPGAdmin)
+
+			return nil
+		}
+
+		err := r.reconcilePGAdminUsers(ctx, pgadmin)
+		assert.ErrorContains(t, err, "strconv.Atoi: parsing")
```

Comment:
:thinking: What error is this catching? I'm not quite tracking what's going on at this point.

Reply:
I've since changed this test and removed this assertion, but it *was* checking that the error returned from `reconcilePGAdminUsers` was a `strconv` parsing error because the commands that get run in the PodExec weren't retrieving the pgAdmin version because they aren't actually being run in a pgAdmin pod in the tests, so an empty string would get passed to the `strconv.Atoi` resulting in a parsing error... You'll see that in my more recent changes that I force a "fake" version into the `stdout`.